### ISSUE TITLE
unknown data loss issue

### ISSUE THREAD

**Issue by psifertex** [outline/outline#5830]
Recently experienced some serious data loss with zero client-side warning and nothing obvious in the server logs. Was merely working on a draft blog post for a few hours, at one point linked a co-worker and shortly after he viewed the page I realized much of my recent changes were gone.

**To Reproduce**
Unsure of exact reproduction steps. I realize this makes fixes incredibly hard, but want to at least put this here in case others experience similar issues.

**Expected behavior**
I would _very much_ like to not lose data. :-( 

**Additional Info**
Here's some data from the server logs in case it's relevant for that particular page.
[jordan-wiki-bug_2023-09-14.log.zip](https://github.com/outline/outline/files/12610801/jordan-wiki-bug_2023-09-14.log.zip)

**Outline:**
- Install: Self Hosted
- Version: v0.71.0 (this is from the /settings/details page, if not sufficiently detailed, let me know)

**Desktop (please complete the following information):**
 - OS: MacOS
 - Browser: Chrome
 - Version: 116.0.5845.179

FWIW, our [org](https://github.com/Vector35) has been a sponsor for a little while now if it helps.

**Comment by tommoor** [comment]
It's good to see so many logs, but nothing really jumps out in terms of being an error.

I assume the post you were working on is "Coruscant Release Notes" – it looks like 12 revisions were created of this document, so there should be many restore points available under "History", it might help point to what happened?

**Comment by psifertex** [comment]
Unfortunately it doesn't. There was only a single checkpoint for the whole day at that point and it only had a small fraction of the changes in it. I assume it was some client side bug but not sure if there's any client side logging that might indicate a problem?

**Comment by tommoor** [comment]
Yea, I misread the revisions are for another document ID.
If you load the same document now are changes saving correctly? aka it was temporary in the tab for that amount of time.

Short of you still having the tab open and there being errors in the developer console there's no additional logging once the tab is closed.

**Comment by psifertex** [comment]
I had two tabs open but closed one. I will check the console of the other.

**Comment by routerino** [comment]
Is there a situation where the client can continue to input data without a websocket connection? If so, it probably should have a small banner saying "lost connection, changes will be synced when reconnected", and enabling the browser warning about "closing this page may lose data".

**Comment by tommoor** [comment]
Editing without a websocket connection has always been supported, changes are stored in IndexedDB in the browser and synced when the connection returns. While clearly something went awry here, generally going offline and even closing the browser tab should not be a worry for losing data.

**Comment by zeekhuge** [comment]
+1 
I too have encountered some (multiple) random data-loss events. But even with significant efforts to be able to reproduce them, have been unable to.

In general, the way data loss happens is the same as mentioned above i.e. recent edits do not appear in the documents.

**Comment by zeekhuge** [comment]
Maybe this is related to [the following discussion](https://github.com/outline/outline/discussions/5791).

**Comment by JustDoItSascha** [comment]
@zeekhuge it's 100% the same issue.

And I think it's obvious what happens here. Changes in a document are not getting merged, but always synced as a whole. When coworker is offline and I change something it's saved. But when coworker comes online (and maybe has changed one letter), his Version overrides my changes because they are newer. There needs to be a merging strategy. Or at least, as a Quick fix, a Version checking and a warning that there is a merge conflict.

Disclaimer: I don't know if that's what happens, it's just an educated guess until someone says that there IS a merging strategy and I'm wrong...

**Comment by tommoor** [comment]
> it's just an educated guess until someone says that there IS a merging strategy and I'm wrong...

I already [mentioned above](https://github.com/outline/outline/issues/5830#issuecomment-1721608200), all document changes are stored and merged with a CRDT. In theory there is full support for unlimited offline editing and the changes will merge upon connection.

**Comment by JustDoItSascha** [comment]
Hm, ok I don't know what CRDT means. But that there is full support for offline editing was not questioned. But there is a problem with syncing when there were changes after I went offline...

**Comment by zeekhuge** [comment]
@tommoor This happened again.

Here's what I did:

1. I opened a doc, enabled Edit mode, and was just moving up & down the doc. Pressed `Esc` and moved to a different doc.
2. I received an email, that this particular doc was edited.
3. I went back, opened up the history and saw an entire `Tip` section was removed.
4. Clicked on the previous version and was trying to `Restore` that version, that did not work either.

Weirdly enough, I have observed loss of data in those highlighted sections (`Tip`, `Warning`, `Info`, etc) some time ago as well.

How do I provide you with relevant info/logs for this, to be able to debug better ?

**Comment by sncat** [comment]
I am not sure if it is exactly the same, but we are experiencing the same phenomenon in our environment.
(This is the self-hosted version of Ver 0.73.1)

One user inquired about the loss of a document that was edited by multiple people.
The edited data of the lost part is not shown in the data change history, and rollback is not possible.

**Comment by ruimelodev** [comment]
@tommoor This is still happening.

Outline: v1.3.0
Browser: Google Chrome Version 144.0.7559.60

I have experienced a severe data loss event on a self-hosted instance. After working extensively on a large document for 7 days, the document suddenly reverted to a version from 5 days ago. The revisions table in the database confirms that no intermediate saves were persisted, despite the UI showing the content while I was editing.

Sequence of Events:
1. I have been editing a single large document for the past 7 days (thousands of lines, multiple attachments/images).
2. I was working intensively on it for the last 3 days.
3. Yesterday, while making minor changes, my session expired and I was logged out.
4. Upon logging back in immediately, the document content had reverted to the state it was in 5 days ago.
5. The "History" UI only shows the version from 5 days ago; all recent history is missing.
  

I have attempted to recover the data from the following sources, all with negative results:
- Database (documents table): The current text reflects the old version.
- Database (revisions table): There are no rows for the missing 5-day period. It appears the saves were never committed to the DB.
- Browser Storage: I checked Chrome DevTools (IndexedDB, LocalStorage, and Cache) but could not find the orphaned state.

It appears to be a silent failure of WebSocket synchronization. The client allowed me to continue writing for days without successfully persisting data to the backend, and provided no UI warning that the connection or save state was failing. When the session expired, the local state was discarded in favor of the last known server state.
